[DOC] Adding user-guide docs for CI testing and constraint-based discovery (#126)
py-why/dodiscover: doc/conditional_independence.rst

.. _conditional_independence:

========================

> Conditional Independence

Reviewer: I am wondering if simply "Independence" would be better. It currently sounds like that the unconditional case is a special case of the conditional independence simply with the conditioning set being empty. Although you can definitely see it like that, I would probably still distinguish between them more clearly. Why not just introduce independence tests in general and mention that one can either perform a unconditional or conditional test?

Author: Done. Lmk wdyt.

========================

.. currentmodule:: dodiscover.ci

Testing for invariances in the data, such as conditional independence (CI) can be represented graphically
in the form of d-separation statements under the causal faithfulness assumption. Given this, one

> is interested in high-powered and well-controlled CI tests that can be used to test for CI in data.

Reviewer: While it is a good fit for dodiscover, I wouldn't make any causal or general graph reference at this point, since (conditional) independence tests are just hypothesis tests as mentioned in the next paragraph. However, it fits to the dodiscover story of course, but if we move it to the stats package, this part should probably be re-written/removed.

Maybe you can move this graphical remark to the constrained based causal discovery part?

Author: Done. Lmk wdyt.

CI tests are framed as a statistical hypothesis test, with the following null hypothesis for a given
pair of variables ``(X, Y)`` and a conditioning set ``Z`` (which may be empty):

> :math:`H_0: X \perp Y | Z`, or written in terms of their distribution :math:`H_0: P(Y | X, Z) = P(Y | Z)`

Reviewer: Write it out as a text as well: "The null hypothesis is that X and Y are statistically independent given a conditioning set Z, i.e., :math:`H_0: X \perp Y | Z`, or written in terms of their distribution :math:`H_0: P(Y | X, Z) = P(Y | Z)`"

Similarly, the alternative hypothesis is written as:

> :math:`H_0: X \not\perp Y | Z`, or written in terms of their distribution :math:`H_0: P(Y | X, Z) \neq P(Y | Z)`

Reviewer: Shouldn't this be H_1 then? One should also emphasize that it is impossible to reject this in practice and, thus, we can only reject independence.

Author: Yes good catch fixed.

> Then typically, one posits an acceptable Type I error rate (false positive), typically :math:`\alpha = 0.05`

Reviewer 2: nit: "acceptable _upper bound_ on the type I error rate"

> Then typically, one posits an acceptable Type I error rate (false positive), typically :math:`\alpha = 0.05`

Author: ```suggestion
Then typically, one posits an acceptable upper bound on the Type I error rate (false positive), typically :math:`\alpha = 0.05`
```

and then either attempts to sample from the null distribution, or characterizes the asymptotic distribution
of the test statistic. In both approaches a pvalue is computed, which is compared to :math:`\alpha`. The pvalue

> states the "probability that we observe our data (e.g. test statistic) under the null hypothesis". By rejecting

Reviewer 2: "probability of observing a t-statistic at _least as extreme_ as our observed t-statistic in null distribution"

> states the "probability that we observe our data (e.g. test statistic) under the null hypothesis". By rejecting

Author: ```suggestion
states the "probability of observing a test-statistic at least as extreme as our observed test-statistic in null distribution". By rejecting
```

the null hypothesis, one claims that :math:`X \not\perp Y | Z`, so that X and Y are in fact (conditionally)
dependent given Z. Note that if one fails to reject the null hypothesis, it is simply by convention that we

> claim X and Y are conditionally independent. It is not necessarily the case, and it is plausible that there

Reviewer: Maybe formulate the "... that we claim X and Y are conditionally independent."-part a bit weaker. Something like "Note that if one fails to reject the null hypothesis, we cannot accept the alternative hypothesis of independence strictly speaking. However, in practice, such as in causal discovery, we would still conclude that they are independent." or something along the line.

Author: Added.

on the underlying distributions. Unfortunately, CMI is notoriously difficult to estimate. There are
various proposals in the literature for estimating CMI, which we summarize here:

- The Kraskov, Stogbauer and Grassberger (KSG) estimate approach estimates mutual information
  via k-NN statistics :footcite:`kraskov_estimating_2004`. It was generalized to CMI by :footcite:`frenzel_partial_2007`.
  This class of estimators is asymptotically correct, meaning if we had an infinite amount of data
  we would obtain the true value of the CMI. However, it relies on statistics generated from the k-NN,
  which if we implement the naive approach using a KDTree, then it generally suffers in high-dimensions.
  In our examples, we see it suffer with dimensionality > 4 or 5.

  Estimates of CMI can be converted into a CI hypothesis test by permutation testing :footcite:`Runge2018cmi`.
  One can generate estimated samples from the null distribution by permuting samples in an intelligent manner
  and then the CMI value generated from the actual observed data can be compared to the CMI values computed
  on the permutated datasets to estimate a pvalue.

>   It is worth noting that if one has "robust" estimates of k-NN using for example a model that

Reviewer: What does "robust estimates of k-NN" mean? I guess you mean the quantity that you estimate using k-NN (e.g., mutual information), maybe just focus on that. Otherwise, k-NN would simply tell you the k neighbors and that just depends on your distance measure.

Author: Sorry I reworded it. So the KSG estimate of MI is based on kNN-based statistics. But it is a theoretical result that as dimensionality increases, the number of samples needed for kNN algorithm grows exponentially. Thus the MI estimate has at least the same if not worse behavior. Assuming you have an alternative algorithm for estimating nearest-neighbors that is robust to noise, then maybe the KSG estimate of MI performs better.

  is effective in high-dimensions, then the KSG estimator for CMI may be quite good. For example,
  one can use variants of Random Forests to generate adaptive nearest-neighbor estimates in high-dimensions

>   or on manifolds, such that the KSG estimator is still quite good.

Reviewer: I would remove the "quite good"s here. Just say that it can be 'powerful' even in high dimensions if the knn estimator can capture it.

Author: Done.

.. autosummary::
   :toctree: generated/

    CMITest

- The Classifier Divergence approach estimates CMI using a classification model.

.. autosummary::
   :toctree: generated/

    ClassifierCMITest

- Direct posterior estimates can be implemented with a classification model by directly
  estimating :math:`P(y|x)` and :math:`P(y|x,z)`, which can be used as plug-in estimates
  to the equation for CMI.

Partial (Pearson) Correlation
-----------------------------
Partial correlation based on the Pearson correlation is equivalent to CMI in the setting
of normally distributed data. Computing partial correlation is fast and efficient and

> thus attractive to use. However, this **relies on the assumption of multivariate Gaussianity**,

Reviewer 2: "Gaussianity"?

Author: Seems like it is a word? https://en.wikipedia.org/wiki/Non-Gaussianity hehe. Any suggestions?

Reviewer: I think Gaussianity is quite common, but the combination with "multivariate" sounds odd. What about "However, this relies on the assumption that the variables are Gaussian."

Author: Modified

which may be unrealistic in certain datasets.

.. autosummary::
   :toctree: generated/

    FisherZCITest

Discrete, Categorical and Binary Data
-------------------------------------
If one has discrete data, then the test to use is based on Chi-square tests. The :math:`G^2`
class of tests will construct a contingency table based on the number of levels across
each discrete variable. An exponential amount of data is needed for increasing levels
for a discrete variable.

.. autosummary::
   :toctree: generated/

    GSquareCITest

Kernel-Approaches
-----------------
Kernel-based tests are attractive since they are semi-parametric and use kernel-based ideas
that have been shown to be robust in the machine-learning field. The Kernel CI test is a test
that computes a test statistic from kernels of the data and uses permutation testing to
generate samples from the null distribution :footcite:`Zhang2011`, which are then used to
estimate a pvalue.

.. autosummary::
   :toctree: generated/

    KernelCITest

Classifier-based Approaches
---------------------------
Another suite of approaches that rely on permutation testing is the classifier-based approach.

> By shuffling data in an intelligent way, one can setup a hypothesis test for CI based on the

Reviewer: What does this "intelligent" mean here? Wouldn't you just uniformly shuffle them in practice?

Author: Marginal ind testing = shuffling the data unifromly
CI testing = shuffling the data conditioned on Z (done via kNN). This is a subroutine in any permutation type procedure.

Author: I reworded this section tho

predicted probabilities from a classification-model. Intuitively, if the shuffled data is similar 
to the unshuffled data, such that the classification-model achieves non-trivial performance
(e.g. >50\% accuracy on a balanced dataset), then one fails to reject the null hypothesis and would
state that the original data was in fact CI :footcite:`Sen2017model`.

.. autosummary::
   :toctree: generated/

>     ClassifierCITest

Reviewer: Not relevant for this PR: But why isn't there one based on regression? So, imply checking whether the model improves when you add the other variable(s)?

Author: I just never implemented one because I saw dowhy had it. We can consolidate now that we moving CI stuff to pywhy-stats :). I think upon skimming, the meta-method for helping generate the pvalue can be consolidated with a suite of meta-methods for pvalue estimation.
